internal/controller/managedcluster_controller.go

```diff
@@ -683,6 +695,357 @@ func (r *ManagedClusterReconciler) machinesAvailable(ctx context.Context, namesp
+		"cloud-config": azureJSON,
+	}
+
+	return makeSecret("azure-cloud-provider", "kube-system", secretData), nil
+}
+
+func (r *ManagedClusterReconciler) propagateVSphereSecrets(ctx context.Context, managedCluster *hmc.ManagedCluster, kubeconfSecret *corev1.Secret) error {
+	vsphereCluster := &capv.VSphereCluster{}
+	if err := r.Client.Get(ctx, client.ObjectKey{
+		Name:      managedCluster.Name,
+		Namespace: managedCluster.Namespace,
+	}, vsphereCluster); err != nil {
+		return fmt.Errorf("failed to get VSphereCluster %s: %s", managedCluster.Name, err)
+	}
+
+	vsphereClIdty := &capv.VSphereClusterIdentity{}
+	if err := r.Client.Get(ctx, client.ObjectKey{
+		Name: vsphereCluster.Spec.IdentityRef.Name,
+	}, vsphereClIdty); err != nil {
+		return fmt.Errorf("failed to get VSphereClusterIdentity %s: %s", vsphereCluster.Spec.IdentityRef.Name, err)
+	}
+
+	vsphereSecret := &corev1.Secret{}
+	if err := r.Client.Get(ctx, client.ObjectKey{
+		Name:      vsphereClIdty.Spec.SecretName,
+		Namespace: r.SystemNamespace,
+	}, vsphereSecret); err != nil {
+		return fmt.Errorf("failed to get VSphere Secret %s: %s", vsphereClIdty.Spec.SecretName, err)
+	}
+
+	vsphereMachines := &capv.VSphereMachineList{}
+	if err := r.Client.List(
+		ctx,
+		vsphereMachines,
+		&client.ListOptions{
+			Namespace: managedCluster.Namespace,
+			LabelSelector: labels.SelectorFromSet(map[string]string{
+				hmc.ClusterNameLabelKey: managedCluster.Name,
+			}),
+			Limit: 1,
+		},
+	); err != nil {
+		return fmt.Errorf("failed to list VSphereMachines for cluster %s: %s", managedCluster.Name, err)
+	}
+	ccmSecret, ccmConfig, err := generateVSphereCCMConfigs(vsphereCluster, vsphereSecret, &vsphereMachines.Items[0])
+	if err != nil {
+		return fmt.Errorf("failed to generate VSphere CCM config: %s", err)
+	}
+	csiSecret, err := generateVSphereCSISecret(vsphereCluster, vsphereSecret, &vsphereMachines.Items[0])
+	if err != nil {
+		return fmt.Errorf("failed to generate VSphere CSI secret: %s", err)
+	}
+
+	if err := applyCCMConfigs(ctx, kubeconfSecret, ccmSecret, ccmConfig, csiSecret); err != nil {
+		return fmt.Errorf("failed to apply VSphere CCM/CSI secrets: %s", err)
+	}
+
+	return nil
+}
+
+func generateVSphereCCMConfigs(vCl *capv.VSphereCluster, vScrt *corev1.Secret, vMa *capv.VSphereMachine) (*corev1.Secret, *corev1.ConfigMap, error) {
+	ns := "kube-system"
+	secretName := "vsphere-cloud-secret"
+	secretData := map[string][]byte{
+		fmt.Sprintf("%s.username", vCl.Spec.Server): vScrt.Data["username"],
+		fmt.Sprintf("%s.password", vCl.Spec.Server): vScrt.Data["password"],
+	}
+	ccmCfg := map[string]any{
+		"global": map[string]any{
+			"port":            443,
+			"insecureFlag":    true,
+			"secretName":      secretName,
+			"secretNamespace": ns,
+		},
+		"vcenter": map[string]any{
+			vCl.Spec.Server: map[string]any{
+				"server": vCl.Spec.Server,
+				"datacenters": []string{
+					vMa.Spec.Datacenter,
+				},
+			},
+		},
+		"labels": map[string]any{
```

Comment:
I am not aware of such labels being set, why do we need them and with such values?

Reply:
These are not k8s labels, it's labels for use in multi datacenter or multi vcenter environment to mark VMs or other stuff. Arbitrary values could passed here. 
I took the default ones they put in [the chart](https://github.com/kubernetes/cloud-provider-vsphere/blob/master/charts/vsphere-cpi/values.yaml#L12)